### PR TITLE
Mxnet allclose

### PR DESCRIPTION
**Pull request by drivanov**
## Description ##
- The analog of `numpy.allclose(a, b, rtol=1e-05, atol=1e-08, equal_nan=False)` is implemented as a MxNet operator:
```
    mx.nd.contrib.allclose(a, b, rtol, atol, equal_nan)
```
- For now, besides the unit test `test_allclose_function()`, this method is used only in
```
def assert_almost_equal(a, b, rtol=None, atol=None, names=('a', 'b'), equal_nan=False)
```
where parameter **a** or/and **b** could be defined also as  `mx.nd.array`(s):
```
    Parameters
    ----------
    a : np.ndarray or mx.nd.array
    b : np.ndarray or mx.nd.array
```
- When calling `assert_almost_equal`, no more `asnumpy()` conversion needed. It will be done automatically (in `assert_almost_equal`), if 
     (a) **a** or **b** has no attribute "context" OR 
     (b) these attributes are different.

- The elimination of `asnumpy()` conversions and the use of `mx.nd.contrib.allclose(...)` for `mx.nd.array`'s allows to achieve 5-7x speedup for GPU tests that use long arrays.

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

## Comments ##
Most of the changed files, were  changed because of the eliminations of `asnumpy()` conversions which were used when `assert_almost_equal` function is called.

### CODE REVIEW

**Comment by karan6181** [comment]
@mxnet-label-bot add [Operator, Test, pr-awaiting-review]

**Comment by chinakook** [comment]
It this op only support float32?

**Comment by drivanov** [comment]
> It this op only support float32?

Actually, only the tolerances (rtol, atol) are supposed to be float32. The elements of tensors could be any type.

**Comment by drivanov** [comment]
I am not sure that this fail is related to my changes made to this PR. I will make an empty commit for triggering CI.

**Comment by drivanov** [comment]
Failing tests are not related to the changes of this PR. I will launch CI one more time.

**Comment by wkcn** [comment]
The CI may fails and raise an irrelated exception. We can disable the using of mx.nd.contrib.allclose firstly to check it.

**Comment by drivanov** [comment]
As I could see that _test_bulking() is not there and I need to fix it.

**Comment by drivanov** [comment]
I think that it is a good time to merge these changes.

**Comment by wkcn** [comment]
If the flaky test re-triggers, I guess that the reason is that MXNet doesn't support the tensor with shape `()`.
We can modify the shape to `(1,)` .

**Comment by wkcn** [comment]
ONNX shape mismatch bug:
http://jenkins.mxnet-ci.amazon-ml.com/blue/organizations/jenkins/mxnet-validation%2Funix-cpu/detail/PR-14443/22/pipeline/294

**Comment by wkcn** [comment]
Hi, @chinakook @szha
Could you help take a review? Thank you!

**Comment by wkcn** [comment]
Hi, @eric-haibin-lin @reminisce 
Could you please help review this PR?
Thank you!

**Comment by szha** [comment]
@drivanov sorry for the delay. Would you mind rebase and resolve conflicts? Thank you.

**Comment by piyushghai** [comment]
@drivanov Can you rebase with the latest master branch ?

**Comment by wkcn** [comment]
I don't know why the unittest fail. It seems that it got crashed.

File: tests/python/gpu/test_gluon_gpu.py [hunk 1]
```diff
@@ -454,10 +522,7 @@ def get_net(num_ops):
 
     time_per_iteration.value = (time.time() - start) / num_iterations
 
-
-@with_seed()
-@unittest.skip('skippping temporarily, tracked by https://github.com/apache/incubator-mxnet/issues/14970')
```
Comment (mseth10) [inline tests/python/gpu/test_gluon_gpu.py]: Why remove this? Have we solved the flakiness of test_gluon_gpu.test_bulking (#14970) as part of this PR ?

Comment (drivanov) [reply tests/python/gpu/test_gluon_gpu.py]: Sorry, it is my fault: `@unittest.skip(...) should stay with 
```
@with_seed()
@unittest.skip('skippping temporarily, tracked by https://github.com/apache/incubator-mxnet/issues/14970')
def test_bulking_gluon_gpu():
    _test_bulking(_test_bulking_in_process)
```
I will fix it

**Comment by wkcn** [comment]
It seems that the CI gets stuck when test `LpNorm`.

The unittest I ran:
```
.../home/wkcn/proj/mxnet/tests/python/unittest/test_operator.py:2791: FutureWarning: Using a non-tuple sequence for multidimensional indexing is deprecated; use `arr[tuple(seq)]` instead of `arr[seq]`. In the future this will be interpreted as an array index, `arr[np.array(seq)]`, which will result either in an error or a different result.
  assert_allclose(x.asnumpy()[idx], y.asnumpy())
........[INFO] Setting test np/mx/python random seeds, use MXNET_TEST_SEED=1234 to reproduce.
....[20:17:16] src/operator/tensor/./.././../common/utils.h:450: MXNET_SAFE_ACCUMULATION=1 is recommended for LpNorm with float16 inputs. See https://mxnet.incubator.apache.org/versions/master/faq/env_var.html for more details.
.


^B^[[5~^B^[[5~

```

**Comment by wkcn** [comment]
Glad to see the success of CI : )

@chinakook @szha @mseth10
Could you help take a review? Thank you!

**Comment by abhinavs95** [comment]
@drivanov Could you do a rebase to resolve the merge conflicts?

**Comment by ChaiBapchya** [comment]
@szha @mseth10 Good to merge? Pl review

**Comment by szha** [comment]
there're merge conflicts that need to be resolved first.

**Comment by drivanov** [comment]
Just in case, I rebased. CI is ok and should be no merge conflicts.

**Comment by wkcn** [comment]
Hi @mseth10 and @szha , the conflict has been resolved.

This PR accelerates MXNet allclose, and it accepts `mx.nd.NDArray` and `np.ndarray` without any conversion.

Could you please help take a review? Thank you so much!

**Pull request merged**